### ISSUE TITLE
Validate path as shiny app

### ISSUE THREAD

**Issue by JosiahParry** [rstudio/shiny#4091]
The `shiny::runApp()` function is very flexible and will run an app from a path e.g. `app.R` `www/index.html` or from a directory that contains a `ui.R` and `server.R` function.

It would be great to have a function to validate that a path is indeed a shiny app without having to run `runApp()` and catch if an error occurs.

**Comment by gadenbuie** [comment]
If this is for a third-party tool to help run Shiny apps, I'd recommend either requiring your users to follow a naming convention established by your tool or to ask users to explicitly name the primary file containing the Shiny app.

Unfortunately, when `shiny::runApp()` is given an `.R` file that purports to produce a Shiny application, it'd be very difficult to confirm that the `.R` file indeed creates a Shiny app without actually running the code.

**Comment by JosiahParry** [comment]
Understood. Do you have a suggestion on how to try validating a shiny app via `shiny::runApp()`? The hard part is that it's blocking. 🤔

**Comment by gadenbuie** [comment]
You could maybe start the app in a background R session with callr? You could also source the app script and inspect the return value. According to the `runApp()` docs it should be a shiny app object or a list with `ui` and `server` components (if it doesn't follow naming conventions).

If you can describe your use case a little more, I might be able to provide some more targeted advice. 😄

**Comment by JosiahParry** [comment]
The goal is to prepare a shiny app for deployment. The directory is zipped, copied, and moved else where. Then from that location `shiny::runApp()` is called supplying only the directory path as that is the most flexible (based on doc). 

However before all of that stuff is done, my goal is to verify that the app is indeed a certified runnable shiny app!

**Comment by JosiahParry** [comment]
```r
is_shiny_app <- function(app_dir, poll_duration = 5000L) {
  run_app <- function(app_dir) shiny::runApp(app_dir, port = httpuv::randomPort())
  rp <- callr::r_bg(run_app, list(app_dir = app_dir))
  .poll_res <- rp$poll_io(poll_duration) # set polling duration
  while (TRUE) {
    line <- rp$read_error_lines()
    if (length(line) > 0) {
      error_detected <- any(grepl("Error in shinyAppDir", line))
      app_start_detected <- any(grepl("Listening on", line))
      if (error_detected || app_start_detected) {
        break
      }
    }
  }

  .killed_process <- rp$kill()

  !error_detected
}
```

**Comment by ismirsehregal** [comment]
Why don't you use `shinyAppDir()`?

it will throw an error if the directory doesn't contain a vaild shiny app, which you could `tryCatch`.

```
result <- shinyAppDir("C:/path/to/shiny/app")
is.shiny.appobj(result)
```

**Comment by ismirsehregal** [comment]
No, `shinyAppDir` doesn't run the app it only creates a shiny app object which you need to `print` or pass to `runApp` to be executed.

**Comment by JosiahParry** [comment]
@ismirsehregal thank you for this hint! This works much better: 

```r
is_shiny_app <- function(path) {
    tryCatch(
      !is.null(shiny::shinyAppDir(path)),
      error = function(cond) FALSE
    )
}
```
